### PR TITLE
federation: Refactoring namespaced resources deletion code from kube ns controller and sharing it with fed ns controller

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/33612
Fixes https://github.com/kubernetes/kubernetes/issues/31975

Refactoring code in kube namespace controller to delete all resources in a namespace when the namespace is deleted. Refactored this code into a separate NamespacedResourcesDeleter class and calling it from federation namespace controller.
This is required for enabling cascading deletion of namespaced resources in federation apiserver.
Before this PR, we were directly deleting the namespaced resources and assuming that they go away immediately. With cascading deletion, we will have to wait for the corresponding controllers to first delete the resources from underlying clusters and then delete the resource from federation control plane. NamespacedResourcesDeleter has this waiting logic.

cc @kubernetes/sig-federation-misc @caesarxuchao @derekwaynecarr @mwielgus

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35950)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by liggitt** [comment]
cc @derekwaynecarr @deads2k

**Comment by nikhiljindal** [comment]
Updated code to fix verify-bazel and verify-test-owners

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 1]
```diff
@@ -487,6 +513,16 @@ func estimateGracefulTerminationForPods(kubeClient clientset.Interface, ns strin
 	return estimate, nil
 }
 
+func unstructuredToPod(obj *runtime.Unstructured) (*api.Pod, error) {
```
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: @caesarxuchao I want to highlight this specially. Before this PR we were using kubeclient to list pods. With this change I am only passing NamespaceInterface here and using the dynamic client to list pods. I cant pass CoreInterface since CoreInterface from federation clientset is different than the CoreInterface in kube clientset (for ex: federation does not have component statuses). So I had 2 options:
* Pass both NamespaceInterface and PodInterface
* Use the dynamic client to list pods.

The second option I think is better. It allows better extensibility when more resources support graceful termination.

Comment (caesarxuchao) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Probably I missed something here. I thought the fed controller also has an instance of kube clientset, can you pass the CoreInterface of that one?

Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: That is only to talk to underlying clusters not federation control plane

Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Btw, the test to verify that pods are deleted when namespace is deleted: https://github.com/kubernetes/kubernetes/blob/2319c020b9a7bedcee68bff97eecf28f2c383366/test/e2e/namespace.go#L80 passed on this PR. So this is working fine. Wanted to make sure that it makes sense. Using dynamic client seems fine to me.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 2]
```diff
@@ -25,66 +25,190 @@ import (
...
 	// assume a default estimate for finalizers to complete when found on items pending deletion.
 	finalizerEstimateSeconds int64 = int64(15)
 )
 
-// operationKey is an entry in a cache.
-type operationKey struct {
-	op  operation
-	gvr unversioned.GroupVersionResource
+// OperationKey is an entry in a cache.
+type OperationKey struct {
```
Comment (smarterclayton) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Why is this public?

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 3]
```diff
@@ -25,66 +25,190 @@ import (
...
 	"k8s.io/kubernetes/pkg/client/typed/dynamic"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/sets"
 
 	"github.com/golang/glog"
 )
 
-// contentRemainingError is used to inform the caller that content is not fully removed from the namespace
-type contentRemainingError struct {
+func NewNamespacedResourcesDeleter(nsClient internalversion.NamespaceInterface,
```
Comment (smarterclayton) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Return an interface.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 4]
```diff
@@ -25,66 +25,190 @@ import (
...
+	err := d.nsClient.Delete(namespace.Name, opts)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+// ResourcesRemainingError is used to inform the caller that all resources are not yet fully removed from the namespace.
+type ResourcesRemainingError struct {
 	Estimate int64
```
Comment (smarterclayton) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Godoc now that this is public

File: pkg/controller/namespace/namespace_controller.go [hunk 5]
```diff
@@ -70,23 +73,24 @@ func NewNamespaceController(
 	// we found in practice though that some auth engines when encountering paths they don't know about may return a 50x.
 	// until we have verbs, we pre-populate resources that do not support list or delete for well-known apis rather than
 	// probing the server once in order to be told no.
-	opCache := operationNotSupportedCache{}
+	opCache := deletion.OperationNotSupportedCache{}
```
Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: Make this an interface or something opaque that has an `Ignore(unversioned.GroupVersionResource)` method that you call to below and have deletion check `IsIgnored(gvr)`

File: pkg/controller/client/client.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+package client
+
+import (
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+)
+
+// Instantiates a new dynamic client pool with the given config and user agent.
+func NewDynamicClientPool(cfg *restclient.Config) dynamic.ClientPool {
```
Comment (derekwaynecarr) [inline pkg/controller/client/client.go]: I must be missing something, but why is this needed?

Why is it in this package and not in the existing `pkg/client/typed/dynamic` package?

There is nothing obviously controller specific to me here.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 3 above)
Comment (derekwaynecarr) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: +1

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 7]
```diff
@@ -25,66 +25,190 @@ import (
...
+	clientPool dynamic.ClientPool
+	// Cache of what operations are not supported on each group version resource.
+	opCache OperationNotSupportedCache
+	// Group version resources that should be operated on.
+	groupVersionResources []unversioned.GroupVersionResource
+	// The finalizer token that should be removed from the namespace
+	// when all resources in that namespace have been deleted.
+	finalizerToken api.FinalizerName
+	// Also delete the namespace when all resources in the namespace have been deleted.
+	deleteNamespaceWhenDone bool
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Just leaving a note here on what the use case is to not also delete the namespace when done...

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 8]
```diff
@@ -25,66 +25,190 @@ import (
...
 }
 
-func (e *contentRemainingError) Error() string {
+func (e *ResourcesRemainingError) Error() string {
 	return fmt.Sprintf("some content remains in the namespace, estimate %d seconds before it is removed", e.Estimate)
 }
 
 // operation is used for caching if an operation is supported on a dynamic client.
-type operation string
+type Operation string
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: i think this can stay private... it should all be internal to the namespace deleter...

File: pkg/controller/namespace/namespace_controller.go [hunk 9]
```diff
@@ -50,9 +51,11 @@ type NamespaceController struct {
 	// list of preferred group versions and their corresponding resource set for namespace deletion
 	groupVersionResources []unversioned.GroupVersionResource
 	// opCache is a cache to remember if a particular operation is not supported to aid dynamic client.
-	opCache operationNotSupportedCache
+	opCache deletion.OperationNotSupportedCache
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: shouldn't this now be self contained to the namespacedResourcesDeleter?

File: pkg/controller/namespace/namespace_controller.go (hunk 5 above)
Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: I would prefer we keep opCache entirely as an internal piece of logic to the deleter object you defined and not have it at this scope.

File: pkg/controller/namespace/namespace_controller.go [hunk 10]
```diff
@@ -70,23 +73,24 @@ func NewNamespaceController(
...
 	// probing the server once in order to be told no.
-	opCache := operationNotSupportedCache{}
+	opCache := deletion.OperationNotSupportedCache{}
 	ignoredGroupVersionResources := []unversioned.GroupVersionResource{
 		{Group: "", Version: "v1", Resource: "bindings"},
 	}
 	for _, ignoredGroupVersionResource := range ignoredGroupVersionResources {
-		opCache[operationKey{op: operationDeleteCollection, gvr: ignoredGroupVersionResource}] = true
-		opCache[operationKey{op: operationList, gvr: ignoredGroupVersionResource}] = true
+		opCache[deletion.OperationKey{Operation: deletion.OperationDeleteCollection, Gvr: ignoredGroupVersionResource}] = true
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: i would move this into the namespace deleter

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]

**Comment by nikhiljindal** [comment]
Am working on rebasing and comments

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Added back PodInterface

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Reverted to being private

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 11]
```diff
@@ -25,66 +25,190 @@ import (
...
+	Operation Operation
+	Gvr       unversioned.GroupVersionResource
 }
 
-// operationNotSupportedCache is a simple cache to remember if an operation is not supported for a resource.
-// if the operationKey maps to true, it means the operation is not supported.
-type operationNotSupportedCache map[operationKey]bool
+// OperationNotSupportedCache is a simple cache to remember if an operation is not supported for a resource.
+// if the OperationKey maps to true, it means the operation is not supported.
+type OperationNotSupportedCache map[OperationKey]bool
```
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Yes. Reverted to being private

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Added an interface

File: pkg/controller/namespace/namespace_controller.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/namespace_controller.go]: yes changed to internal

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 8 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: yes done

File: pkg/controller/namespace/namespace_controller.go (hunk 9 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/namespace_controller.go]: yes, fixed

File: pkg/controller/namespace/namespace_controller.go (hunk 10 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/namespace_controller.go]: moved

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 7 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: In federation, we can not delete the namespace when all resources in that namespace are deleted.
We also have to first delete the namespace from all underlying clusters and then delete the namespace. Hence, we set this to true in federation.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Sorry did not understand. Is it not a valid GoDoc comment?

**Comment by nikhiljindal** [comment]
Thanks for your comments @smarterclayton @liggitt and @deads2k 

Rebased and updated code as per comments. PTAL.

File: pkg/controller/client/client.go (hunk 6 above)
Comment (nikhiljindal) [inline pkg/controller/client/client.go]: Moved to `pkg/client/typed/dynamic`

**Comment by nikhiljindal** [comment]
Rebased and updated as per comments

**Comment by nikhiljindal** [comment]
All tests passed! :)

**Comment by nikhiljindal** [comment]
I have already rebased multiple times. Will wait for you guys to take a look before rebasing.
Most of the rebases are simple triggered by packages moving out to client-go tools or machinery or util repo.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 12]
```diff
@@ -80,23 +252,24 @@ func (o *operationNotSupportedCache) setNotSupported(key operationKey) {
...
+		result, err = fn(latestNamespace)
 		if err == nil {
 			return result, nil
 		}
 		if !errors.IsConflict(err) {
 			return nil, err
 		}
 		prevNamespace := latestNamespace
-		latestNamespace, err = kubeClient.Core().Namespaces().Get(latestNamespace.Name, metav1.GetOptions{})
+		//		latestNamespace, err = kubeClient.Core().Namespaces().Get(latestNamespace.Name, metav1.GetOptions{})
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: why leave the commented line?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
just a few more questions

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 13]
```diff
@@ -29,17 +30,188 @@ import (
...
+		deleteNamespaceWhenDone: deleteNamespaceWhenDone,
+	}
+	d.initOpCache()
+	return d
+}
+
+var _ NamespacedResourcesDeleterInterface = &NamespacedResourcesDeleter{}
+
+// NamespacedResourcesDeleter is used to delete all resources in a given namespace.
+type NamespacedResourcesDeleter struct {
```
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Doesnt need to be public anymore. Fixed

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go (hunk 12 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Unintended. Removed

**Comment by nikhiljindal** [comment]
Thanks @derekwaynecarr 
Fixed comments and rebased

**Comment by derekwaynecarr** [comment]
the last commit looks fine to me, squash the commits, and then i will tag.

**Comment by nikhiljindal** [comment]
Thanks @derekwaynecarr !
Squashed the last commit. Let me know if you want me to squash more.

**Comment by nikhiljindal** [comment]
@k8s-bot verify test this

**Review by derekwaynecarr — APPROVED** [review]

**Comment by nikhiljindal** [comment]
`pull-kubernetes-verify` failed again.
@k8s-bot verify test this

File: pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go [hunk 14]
```diff
@@ -306,13 +318,13 @@ func testResources() []*metav1.APIResourceList {
 					Name:       "pods",
 					Namespaced: true,
 					Kind:       "Pod",
-					Verbs:      []string{"get", "list", "delete", "deletecollection", "create", "update"},
+					Verbs:      []string{"get", "list", "delete", "deleteCollection", "create", "update"},
```
Comment (liggitt) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go]: why mixed case for `deleteCollection`?

Comment (liggitt) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go]: @sttts this should be lowercase... policy treats it as lowercase at https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go#L210 and discovery writes it as lowercase at https://github.com/kubernetes/kubernetes/tree/master/pkg/genericapiserver/endpoints/installer.go#L72

File: federation/pkg/federation-controller/namespace/namespace_controller_test.go [hunk 15]
```diff
@@ -153,10 +132,12 @@ func TestNamespaceController(t *testing.T) {
 	ns1.DeletionTimestamp = &metav1.Time{Time: time.Now()}
 	namespaceWatch.Modify(&ns1)
 	assert.Equal(t, ns1.Name, GetStringFromChan(nsDeleteChan))
-	assert.Equal(t, "all", GetStringFromChan(rsDeleteChan))
```
Comment (liggitt) [inline federation/pkg/federation-controller/namespace/namespace_controller_test.go]: are we losing tests here?

Comment (derekwaynecarr) [inline federation/pkg/federation-controller/namespace/namespace_controller_test.go]: i was ok with this change since the code is now leveraging the shared deletion logic which itself has tests to verify the right set of operations were invoked.  duplicating that seemed unnecessary.  maybe a future test could just verify that the deleter itself was invoked?

File: staging/src/k8s.io/client-go/dynamic/client_pool.go [hunk 16]
```diff
@@ -70,6 +71,13 @@ func NewClientPool(config *restclient.Config, mapper meta.RESTMapper, apiPathRes
 	}
 }
 
+// Instantiates a new dynamic client pool with the given config.
+func NewDynamicClientPool(cfg *restclient.Config) ClientPool {
```
Comment (nikhiljindal) [inline staging/src/k8s.io/client-go/dynamic/client_pool.go]: This is causing `hack/verify-staging-imports.sh` to fail with:
```
I0131 11:53:14.800] client-go has a forbidden dependency:
I0131 11:53:14.800] 
I0131 11:53:14.802]   k8s.io/client-go/dynamic imports k8s.io/kubernetes/pkg/api
I0131 11:53:14.802] 
I0131 11:53:14.803] FAILED   hack/make-rules/../../hack/verify-staging-imports.sh
```

So I will have to move this elsewhere.
I had it at `pkg/controller/client` earlier.
@derekwaynecarr Is that fine or is there some other place you want me to put this at?

File: staging/src/k8s.io/client-go/dynamic/client_pool.go [hunk 17]
```diff
@@ -70,6 +71,13 @@ func NewClientPool(config *restclient.Config, mapper meta.RESTMapper, apiPathRes
...
+	// TODO: should use a dynamic RESTMapper built from the discovery results.
+	restMapper := api.Registry.RESTMapper()
```
Comment (sttts) [inline staging/src/k8s.io/client-go/dynamic/client_pool.go]: There is also a registry for client-go. Note that there is a difference between client and kube types. Kube's api.Registry is relevant only on the apiserver side. If you talk to the apiserver via a client most probably you mean client-go's registry here.

Comment (sttts) [inline staging/src/k8s.io/client-go/dynamic/client_pool.go]: `client-go/pkg/api.Registry`

File: pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go (hunk 14 above)
Comment (nikhiljindal) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go]: Thanks for pointing that out @liggitt. Fixed to lowercase.

File: staging/src/k8s.io/client-go/dynamic/client_pool.go (hunk 17 above)
Comment (nikhiljindal) [inline staging/src/k8s.io/client-go/dynamic/client_pool.go]: Thanks, updated to use `client-go/pkg/api.Registry`

**Comment by nikhiljindal** [comment]
Thanks @liggitt, @sttts and @derekwaynecarr for your comments. Updated the code.

**Comment by nikhiljindal** [comment]
Rebased

**Comment by nikhiljindal** [comment]
@k8s-bot cvm gce e2e test this

**Comment by nikhiljindal** [comment]
Assigning to @brendandburns for OWNERS approval

**Comment by nikhiljindal** [comment]
Rebased

**Comment by nikhiljindal** [comment]
Adding back LGTM and approved labels after trivial rebase

**Comment by nikhiljindal** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by nikhiljindal** [comment]
the deployment test is really flaky. It has failed twice consecutively now. Ref https://github.com/kubernetes/kubernetes/issues/39785.

	@k8s-bot gce etcd3 e2e test this

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this
